### DISCUSSION TITLE
connecting to A K40 laser (designing my own driver board)

### DISCUSSION THREAD

**Discussion by EmbeddedHobbyist** [fra589/grbl-Mega-5X#341] [Q&A]
Hi I'm the embedded hobbyist, which is just a nick name and the name of my YouTube channel :-)

I'm currently looking at upgrading my K40 laser so i can use LightBurn to control it and so i need to replace the controller board it came with. I had a Mega laying around so thought grbl-mega might be the way to go, But grbl-mega-5x seems to be more active and could provide a better upgrade path.

So here's my problem,

I want to get my head around the pins i need to connect to when designing the driver board to fit the Mega.
I need the following signals

1: Laser PWM to provide the power to the "IN" input on the K40 power supply to control laser power
2: Laser fire to provide the enable to the "Fire" input on the PSU to enable the laser.

With just the bare Mega i can use the command
$32=1
M3 G1 X100 S500 F1000

the laser PWM starts and shows about 50% duty cycle on the scope
But i cant find the signal i need to enable the laser.

Am i correct in thinking that the "spindle enable" should be active during the above command?

Can somebody point me in the right direction?

Thanks in Advance
Teh

**Comment by fra589** [comment]
Hi @EmbeddedHobbyist,  
  
You can find the grbl-Mega-5X pinout in the wiki: https://github.com/fra589/grbl-Mega-5X/wiki/grbl-Mega-5X-pinout

By default, the PWM output is on the pin D8.

The laser is driven by the spindle command: M3 & M4 for laser on and M5 for power off laser (the difference between M3 & M4 is just the status of the spindle direction pin (D5 by default) which is generaly not used by laser).

S\<number\> is driving the PWM power. Proportional to \<number\>:
- If \<number\> >= SPINDLE_PWM_MAX_VALUE ($35 parameter) this output PWM to 100% duty cycle.
- if \<number\> <= SPINDLE_PWM_MIN_VALUE ($36 parameter) this sets PWM to 0.4% or lowest duty cycle.
- If \<number\> = 0 this disable the PWM output.

When the GCode status of spindle is M3 or M4 and the spindle S value is strictly upper than zero, the SPINDLE_ENABLE pin (default to D4) is set to up. In any other cases (M5 or S0) the SPINDLE_ENABLE pin is set to down.

@++;
Gauthier.

**Reply by EmbeddedHobbyist** [reply]
Hi Gauthier, thanks for getting back so quick.

I could not get the spindle_enable signal to go high with 
$32=1 Laser Mode
S=5000 Mid RPM speed point
M3 G1 X100 F1000
 or
M4 G1 X100 F1000

The spindle direction bit would toggle between M3 and M4
M3 would leave the PWM at the setting value

M4 would ramp up and hold setting point then ramp down.

I could not get any action uut of the spindle_enable at any point.

Tried the default settings below but that stopped the PWM on D8?

#ifdef SEPARATE_SPINDLE_LASER_PIN
  // Laser PWM can be on D6 (default) or on D8 or D9. 
  //#define LASER_PWM_ON_D6
  #define LASER_PWM_ON_D8
  //#define LASER_PWM_ON_D9
#endif

Not sure if this is the fix but i uncomment the following and the spindle enable started to toggle!
#define SPINDLE_ENABLE_OFF_WITH_ZERO_SPEED // Default disabled. Uncomment to enable.

I now seem to be getting all the signals i need to drive the K40 when the PCB comes back from PCBWay I'll get to try it on it.
Just for the fun of it i'm designing my own driver board,

Regards
TeH

**Reply by fra589** [reply]
Hi TeH,  
  
> I could not get any action uut of the spindle_enable at any point.

?!?! Strange... Should work... I need to verify on develop board. But this will take time since I leave tomorrow for a series of international model aircraft competitions until the end of September.

I added the `#define SEPARATE_SPINDLE_LASER_PIN ` option for hybrid machine which have both spindle and laser.
When using it, the pin for laser must be different than the spindle pin. If this option is active, if $32=0, M3S* will output PWM on the spindle pin, and if $32=1, M3S* will output PWM on the laser pin. In this case, when $31=1, the PWM cycle will be generated between LASER_PWM_MAX_VALUE and LASER_PWM_MIN_VALUE and when $31=0, between SPINDLE_PWM_MAX_VALUE and SPINDLE_PWM_MIN_VALUE on the spindle pin.

> Not sure if this is the fix but i uncomment the following and the spindle enable started to toggle!
> #define SPINDLE_ENABLE_OFF_WITH_ZERO_SPEED // Default disabled. Uncomment to enable.

I need to verify this (end september :-$) perhaps may be a bug on this side.

@++;
Gauthier.
